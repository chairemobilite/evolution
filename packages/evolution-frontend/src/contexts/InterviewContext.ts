/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import * as React from 'react';

// TODO Can this be InterviewResponses from 'evolution-common/lib/services/interviews/interview'? To be confirmed
export type InterviewResponses = { [key: string]: any };

export type InterviewState =
    | { status: 'list'; responses: InterviewResponses }
    | { status: 'creating'; responses: InterviewResponses; username: string }
    | { status: 'error'; responses: InterviewResponses }
    | { status: 'success'; responses: InterviewResponses; interviewUuid: string }
    | { status: 'entering'; responses: InterviewResponses };

type InterviewAction =
    | { type: 'createNew'; username: string }
    | { type: 'list' }
    | { type: 'update_responses'; responses: InterviewResponses }
    | { type: 'success'; interviewUuid: string }
    | { type: 'enter'; queryData: URLSearchParams };

export function interviewReducer(state: InterviewState, action: InterviewAction): InterviewState {
    switch (action.type) {
    case 'createNew':
        return { status: 'creating', responses: state.responses, username: action.username };
    case 'list':
        return { ...state, status: 'list' };
    case 'success':
        return { status: 'success', responses: state.responses, interviewUuid: action.interviewUuid };
    case 'update_responses':
        return { ...state, status: 'list', responses: action.responses };
    case 'enter': {
        // TODO: The source param is hard-coded. If we need more than that, we should add a project config to white list fields
        const source = action.queryData.get('source');
        const alphaNumRegex = /^[a-z0-9-_]+/i;
        const responses: { [key: string]: string } = {};
        if (source) {
            const alphaNumSource = source.match(alphaNumRegex);
            if (alphaNumSource && state.responses.source !== alphaNumSource[0])
                responses.source = alphaNumSource[0];
        }
        return {
            status: 'entering',
            responses:
                    Object.keys(responses).length !== 0
                        ? Object.assign({}, state.responses, responses)
                        : state.responses
        };
    }
    }
}

export const initialState = { status: 'list' as const, responses: {} };

export const InterviewContext = React.createContext<{
    state: InterviewState;
    dispatch: React.Dispatch<InterviewAction>;
}>({
    state: initialState,
    dispatch: () => {
        /* To be replaced with actual dispatch */
    }
});
