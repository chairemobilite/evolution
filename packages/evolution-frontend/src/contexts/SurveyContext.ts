/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import * as React from 'react';

// TODO Properly type this
export type SurveySections = { [key: string]: any };
// TODO This type is probably WidgetConfig from '../services/widgets'. But it is not complete yet and it's to be confirmed
export type SurveyWidgets = { [key: string]: any };

export type SurveyContextType = {
    sections: SurveySections;
    widgets: SurveyWidgets;
    devMode: boolean;
    dispatch: React.Dispatch<SurveyAction>;
};

type SurveyAction = { type: 'setDevMode'; value: boolean };

export function surveyReducer(state: { devMode: boolean }, action: SurveyAction): { devMode: boolean } {
    switch (action.type) {
    case 'setDevMode':
        console.log('Setting dev mode ', action.value);
        return { devMode: action.value };
    }
}

/**
 * This context contains the survey specific data, to build the questionnaire
 * widgets: widget configurations, sections, questions, etc.
 *
 * This data is fixed during the app initialization, is provided by the survey
 * application itself and does not change
 *
 * FIXME: Widgets and sections cannot change for the app. It can go in the
 * application.config instead of context. This way, it will be available even
 * outside react widgets.
 */
export const SurveyContext = React.createContext<SurveyContextType>({
    sections: {},
    widgets: {},
    devMode: false,
    dispatch: () => {
        /* To be replaced with actual dispatch */
    }
});
