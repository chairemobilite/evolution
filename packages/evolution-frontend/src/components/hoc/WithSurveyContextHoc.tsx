/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { SurveyContext, SurveyContextType } from '../../contexts/SurveyContext';

export interface WithSurveyContextProps {
    surveyContext: SurveyContextType;
}

export const withSurveyContext =
    <T extends WithSurveyContextProps = WithSurveyContextProps>(WrappedComponent: React.ComponentType<T>) =>
        (props: Omit<T, keyof WithSurveyContextProps>) =>
            (
                <SurveyContext.Consumer>
                    {(context) => <WrappedComponent {...(props as T)} surveyContext={context} />}
                </SurveyContext.Consumer>
            );
