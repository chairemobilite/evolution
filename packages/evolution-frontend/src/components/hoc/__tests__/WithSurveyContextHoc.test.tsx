/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import TestRenderer from 'react-test-renderer';

import { withSurveyContext, WithSurveyContextProps } from '../WithSurveyContextHoc';
import { SurveyContext, surveyReducer } from '../../../contexts/SurveyContext';

interface TestProps {
    foo: string;
}

export const BaseTestComponent: React.FunctionComponent<{}> = (props: {}) => {
    const [devMode, dispatchSurvey] = React.useReducer(surveyReducer, { devMode: false });
    return (
        <SurveyContext.Provider value={{ sections: { key1: 'a', key2: 2}, widgets: {}, ...devMode, dispatch: dispatchSurvey }}>
            <TestComponentWithContext foo='This is a test component'/>
        </SurveyContext.Provider>
    );
};
  
class TestComponent extends React.Component<TestProps & WithSurveyContextProps> {
    public render() {
        return <div>{Object.keys(this.props.surveyContext.sections)} {this.props.surveyContext.devMode === false ? 'normal' : 'dev' } {this.props.foo}</div>
    }
}

const TestComponentWithContext = withSurveyContext(TestComponent);

test('Test HOC', () => {
    const wrapper = TestRenderer.create(
        <BaseTestComponent/>
    );
    expect(wrapper).toMatchSnapshot();
})