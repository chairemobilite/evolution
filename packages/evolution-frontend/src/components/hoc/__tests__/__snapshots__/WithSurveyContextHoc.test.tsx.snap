// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test HOC 1`] = `
<div>
  key1
  key2
   
  normal
   
  This is a test component
</div>
`;
